ddl/column.go

```diff
@@ -222,14 +222,252 @@ func onAddColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, nil, errors.Trace(err)
+			}
+			logutil.BgLogger().Info("[ddl] run add columns job", zap.String("job", job.String()), zap.Reflect("columnInfo", *columnInfo), zap.Int("offset", offsets[i]))
+
+			// Set offset arg to job.
+			if offset != 0 {
+				offsets[i] = offset
+			}
+			if err = checkAddColumnTooManyColumns(len(tblInfo.Columns)); err != nil {
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, nil, errors.Trace(err)
+			}
+			columnInfos = append(columnInfos, columnInfo)
+		}
+	}
+	job.Args = []interface{}{columnInfos, positions, offsets}
+	return tblInfo, columnInfos, positions, offsets, nil
+}
+
+func setColumnsState(columnInfos []*model.ColumnInfo, state model.SchemaState) {
+	for i := range columnInfos {
+		columnInfos[i].State = state
+	}
+}
+
+func onAddColumns(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	// Handle the rolling back job.
+	if job.IsRollingback() {
+		ver, err = onDropColumns(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	failpoint.Inject("errorBeforeDecodeArgs", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(ver, errors.New("occur an error before decode args"))
+		}
+	})
+
+	tblInfo, columnInfos, positions, offsets, err := checkAddAndCreateColumnInfos(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	originalState := columnInfos[0].State
+	switch columnInfos[0].State {
+	case model.StateNone:
+		// none -> delete only
+		job.SchemaState = model.StateDeleteOnly
+		setColumnsState(columnInfos, model.StateDeleteOnly)
+		ver, err = updateVersionAndTableInfoWithCheck(t, job, tblInfo, originalState != columnInfos[0].State)
+	case model.StateDeleteOnly:
+		// delete only -> write only
+		job.SchemaState = model.StateWriteOnly
+		setColumnsState(columnInfos, model.StateWriteOnly)
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != columnInfos[0].State)
+	case model.StateWriteOnly:
+		// write only -> reorganization
+		job.SchemaState = model.StateWriteReorganization
+		setColumnsState(columnInfos, model.StateWriteReorganization)
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, originalState != columnInfos[0].State)
+	case model.StateWriteReorganization:
+		// reorganization -> public
+		// Adjust table column offsets.
+		oldCols := tblInfo.Columns
+		for i := range offsets {
+			tmpCols := oldCols[:len(oldCols)-len(offsets)+i+1]
+			tblInfo.Columns = tmpCols
+			// For multiple columns with after position, should adjust offsets.
+			// e.g. create table t(a int);
+			// alter table t add column b int after a, add column c int after a;
+			// alter table t add column a1 int after a, add column b1 int after b, add column c1 int after c;
+			// alter table t add column a1 int after a, add column b1 int first;
+			if positions[i].Tp == ast.ColumnPositionAfter {
```

Comment:
Any test for line 344 ~ 350? I can't figure out the reason for those codes, and after remove those line codes, I can still pass all tests.

Reply:
It is adjust offsets for add multi-columns.  For example :
`create table t(a int);`
`alter table t add column (b int, c int);`
`alter table t add column a1 int after a, add column b1 after b;`
So the offset of `a1` is `1`, the offset of `b1` is `2` 
```
a
a1
b1
b
c
```
But the correct order is:
```
a
a1
b
b1
c
```

Actually, I don't test order of the columns, I will add it later.

Reply:
Hi ,I have add some test cases for order of the columns in `ddl/db_integration_test.go`.